### PR TITLE
Add linode provider

### PR DESCRIPTION
**Pull request by guisea**
This is an implementation of a driver for Linode.

### CODE REVIEW

**Comment by guisea** [comment]
As a note.  The "user-data" in this driver is a bash script to do the same job as the Hetzner driver which uses cloud-init.  

I battled for a long time to get this working with cloud-init but there is something funny with Linode config I could not iron out.  As a workaround utilizing their "StackScripts" implementation we can create agents on demand.  

A public "StackScript" I have created allows this to happen from the Linode end.

**Comment by anbraten** [comment]
Thanks for the work. The code looks pretty well already.

I just found this community post regarding cloud-init support: https://www.linode.com/community/questions/22350/cloud-init-support-still-coming (Their staff commented 1day ago 🙈 )

File: providers/linode/driver.go [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+		Region:          d.Region,
+		Type:            d.InstanceType,
+		Label:           agent.Name,
+		Image:           d.Image,
+		StackScriptID:   int(d.StackscriptID),
+		StackScriptData: userdataMap,
+		AuthorizedKeys:  []string{d.SSHKey},
+		RootPass:        d.RootPass,
+		Tags:            d.Tags,
+	})
```
Comment (anbraten) [inline providers/linode/driver.go:141]:
```suggestion
	userDataBase64 := b64.StdEncoding.EncodeToString([]byte(userdataString))

	_, err = d.Client.CreateInstance(ctx, linodego.InstanceCreateOptions{
		Region:         d.Region,
		Type:           d.InstanceType,
		Label:          agent.Name,
		Image:          d.Image,
		AuthorizedKeys: []string{d.SSHKey},
		RootPass:       d.RootPass,
		Tags:           d.Tags,
		Metadata: &linodego.InstanceMetadataOptions{
			UserData: userDataBase64,
		},
	})
```

**Comment by guisea** [comment]
> Thanks for the work. The code looks pretty well already. I just need to test it.
> 
> I just found this community post regarding cloud-init support:
> 
> * https://www.linode.com/community/questions/22350/cloud-init-support-still-coming (Their staff commented 1day ago 🙈 )
> * https://www.linode.com/blog/compute/metadata-service-now-available-in-open-beta/

Haha,  yeah.  That looks promising.  But seems to have been coming for years already.  Being in NZ I deploy resources in the closest GEO which is not in the Beta sadly.  The StackScript/Bash solution works no matter anyway.  I was using for a time this kind of method to get around the shortfall while using Drone Autoscaler.  I did code a linode driver for that too some long time back ;) Never released but after banging ones head re: cloud-init you just give up.

File: go.mod [hunk 2]
```diff
@@ -10,15 +10,17 @@ require (
...
 	golang.org/x/oauth2 v0.8.0
 )
 
 require (
 	github.com/beorn7/perks v1.0.1 // indirect
 	github.com/cespare/xxhash/v2 v2.2.0 // indirect
 	github.com/cpuguy83/go-md2man/v2 v2.0.2 // indirect
+	github.com/go-resty/resty/v2 v2.7.0 // indirect
 	github.com/golang/protobuf v1.5.3 // indirect
+	github.com/linode/linodego v1.20.1 // indirect
```
Comment (anbraten) [inline go.mod:23]: Not sure why its indirect?

**Review by anbraten — COMMENTED** [review]
They added cloud-init to more regions now, so maybe it is already supported in your region as well. Apart from that it lgtm.

**Comment by guisea** [comment]
Not sure myself either is a direct dependency.  I will check out the
cloud-init again in a few days should I get time.

**Comment by anbraten** [comment]
> Not sure myself either is a direct dependency.  I will check out the cloud-init again in a few days should I get time.

Awesome. Sorry for the long silence 🙈 Just poke me if I am missing it again.

**Comment by 6543** [comment]
could you merge master into the feature branch :)

**Comment by runephilosof-karnovgroup** [comment]
Metadata/userdata/cloud-init is accessible to everything running on the agent, not only `root`.
https://www.linode.com/docs/products/compute/compute-instances/guides/metadata/#access-the-metadata-service-api

So you should not save the woodpecker token in that.

I could not find anything in the documentation about how the stackscripts are retrieved by the machine on boot.

**Comment by runephilosof-karnovgroup** [comment]
> So you should not save the woodpecker token in that.

Unless the tokens are one-time use, I haven't looked into that.

File: go.mod (hunk 2 above)
Comment (6543) [reply go.mod:23]: exec `go mod tidy` now ... and it should be fixed

**Comment by 6543** [comment]
https://github.com/woodpecker-ci/autoscaler/pull/15#discussion_r1468502157

**Comment by guisea** [comment]
go mod tidy run against this branch and pushed.  Is now correctly shown as direct dependency

**Review by 6543 — APPROVED** [review]

**Comment by 6543** [comment]
"... This branch has conflicts that must be resolved ..."

Can you merge target branch into pull head branch :)

**Comment by guisea** [comment]
Hows this now?  I have merged changes from main and updated the provider to go.woodpecker-ci.org/woodpecker/v2 
@6543 @anbraten 

Also gave things a quick test on one of my Linode accounts to ensure everything is still in order.  And it was

**Review by anbraten — APPROVED** [review]

File: providers/linode/provider.go [hunk 3]
```diff
@@ -0,0 +1,263 @@
...
+	if err != nil {
+		return fmt.Errorf("%s: RenderUserDataTemplate: %w", d.name, err)
+	}
+	userData := b64.StdEncoding.EncodeToString([]byte(userdataString))
+
+	userdataMap := make(map[string]string)
+
+	var metadata *linodego.InstanceMetadataOptions
+
+	// TODO: remove once linode user-data is out of beta
```
Comment (runephilosof-karnovgroup) [inline providers/linode/provider.go:130]: The user data service was out of beta January 2nd 2024.
https://www.linode.com/blog/compute/customize-deployments-with-akamais-metadata-service/

Comment (guisea) [reply providers/linode/provider.go:130]: It is still not available in all data centres.  For me the closest is ap-southeast in Sydney.  This region does not have metadata service

**Pull request merged**